### PR TITLE
DIOS-6105 Add timestamp checking

### PR DESCRIPTION
**Pull request by harryz2000**
Use rectified timestamp in MediaFrameListenerBridge (for SRT and RTMP) and RTPIncomingMediaStreamDepacketizer (for WebRTC). 

Note a flag was added to MediaFrameListenerBridge to disable the timestamp check by default as the class is also used by transcoder. We would expect the timestamp from transcoder matches the input. So we don't need to check the timestamp again there.

### CODE REVIEW

File: src/MediaFrameListenerBridge.cpp [hunk 1]
```diff
@@ -177,6 +178,29 @@ void MediaFrameListenerBridge::onMediaFrame(DWORD ignored, const MediaFrame& fra
 {
 	timeService.Async([=, frame = std::shared_ptr<MediaFrame>(frame.Clone())] (auto now){
 		
+		if (tsChecker)
+		{
+			// Check timestamp
+			auto [status, rts] = tsChecker->Check(frame->GetTime(), frame->GetTimeStamp(), frame->GetClockRate());
+			if (status != TimestampChecker::CheckResult::Valid)
+			{
+				Error("Invalid timestamp. status: %d, info: %s, corrected: %llu\n", int(status), frame->TimeInfoToString().c_str(), rts);
```
Comment (bcostdolby) [inline src/MediaFrameListenerBridge.cpp:187]: Minor: Can we log the status as a enum string to make the logs easier to read?

File: include/MediaFrameListenerBridge.h [hunk 2]
```diff
@@ -37,7 +38,7 @@ class MediaFrameListenerBridge :
 	};	
 	
 public:
-	MediaFrameListenerBridge(TimeService& timeService, DWORD ssrc, bool smooth = false);
+	MediaFrameListenerBridge(TimeService& timeService, DWORD ssrc, bool smooth = false, bool checkTimestamp = false);
```
Comment (bcostdolby) [inline include/MediaFrameListenerBridge.h:41]: I assume this is to prevent it being used for things like transcode layers? Usage code isnt in this PR

**Review by bcostdolby — COMMENTED** [review]
Should wait until after 2.19.0 to merge IMO

**Comment by bcostdolby** [comment]
Please update the PR title to use the conventional commits format

File: src/MediaFrameListenerBridge.cpp (hunk 1 above)
Comment (harryz2000) [reply src/MediaFrameListenerBridge.cpp:187]: Will do.

File: include/MediaFrameListenerBridge.h (hunk 2 above)
Comment (harryz2000) [reply include/MediaFrameListenerBridge.h:41]: Yes. It is to make it disabled by default

**Comment by harryz2000** [comment]
> Please update the PR title to use the conventional commits format

Have we got one agreed format? :)

**Review by murillo128 — CHANGES_REQUESTED** [review]

File: src/MediaFrameListenerBridge.cpp [hunk 3]
```diff
@@ -177,6 +178,30 @@ void MediaFrameListenerBridge::onMediaFrame(DWORD ignored, const MediaFrame& fra
 {
 	timeService.Async([=, frame = std::shared_ptr<MediaFrame>(frame.Clone())] (auto now){
 		
+		if (tsChecker)
+		{
+			// Check timestamp
+			auto [status, rts] = tsChecker->Check(frame->GetTime(), frame->GetTimeStamp(), frame->GetClockRate());
+			if (status != TimestampChecker::CheckResult::Valid)
+			{
+				Error("Invalid timestamp. status: %d, info: %s, corrected: %llu offset: %lld \n",
```
[the remark this answers is not in the document]
Comment (harryz2000) [reply src/MediaFrameListenerBridge.cpp:187]: The issue here is we don't enable warning level logging. If we change to warning, we would lose the information that is very important for debugging.

File: src/MediaFrameListenerBridge.cpp [hunk 4]
```diff
@@ -177,6 +178,30 @@ void MediaFrameListenerBridge::onMediaFrame(DWORD ignored, const MediaFrame& fra
...
+			if (status != TimestampChecker::CheckResult::Valid)
+			{
+				Error("Invalid timestamp. status: %d, info: %s, corrected: %llu offset: %lld \n", 
+					TimestampChecker::CheckResultToString(status), frame->TimeInfoToString().c_str(), rts, tsChecker->GetTimestampOffset());
+			}
+			
+			if (frame->GetType() == MediaFrame::Video)
+			{
+				VideoFrame* vframe = static_cast<VideoFrame*>(frame.get());
+				auto pts = rts + vframe->GetPresentationTimestamp() - vframe->GetTimestamp();
```
[the remark this answers is not in the document]
Comment (harryz2000) [reply src/MediaFrameListenerBridge.cpp:194]: The rationale is the pts is not far from ts and pts should be always >0. But yes this calculation has issues. I will add another checker for it.

File: src/rtp/RTPIncomingMediaStreamDepacketizer.cpp [hunk 5]
```diff
@@ -40,6 +40,17 @@ void RTPIncomingMediaStreamDepacketizer::onRTP(const RTPIncomingMediaStream* gro
 	 //If we have a new frame
 	 if (frame)
 	 {
+		// Check timestamp
+		auto [status, rts] = tsChecker.Check(frame->GetTime(), frame->GetTimeStamp(), frame->GetClockRate());
```
Comment (murillo128) [inline src/rtp/RTPIncomingMediaStreamDepacketizer.cpp:44]: not sure if i want to put this code here, as it only would change the timestamps for the media frames, but not the timestamps for the rtp packets

Comment (harryz2000) [reply src/rtp/RTPIncomingMediaStreamDepacketizer.cpp:44]: That is right. We don't see a lot of timestamp issues for WebRTC except the rare random invalid timestamps. The check is critical for recordings, A/V sync, etc. But might not be that critical for WebRTC.

But we could change RTP packet timestamp as well.

**Pull request merged**